misc-scripts/xref_mapping/XrefParser/UniProtParser/Transformer.pm

```diff
@@ -0,0 +1,464 @@
+  my ( $direct_xrefs, $dependent_xrefs )
+    = $self->_make_links_from_crossreferences( $accession, $source_id );
+  # Do not assign empty arrays to FOO_XREFS, current insertion code
+  # doesn't like them.
+  if ( scalar @{ $direct_xrefs } > 0 ) {
+    push @{ $xref_graph_node->{'DIRECT_XREFS'} }, @{ $direct_xrefs };
+  }
+  if ( scalar @{ $dependent_xrefs } > 0 ) {
+    push @{ $xref_graph_node->{'DEPENDENT_XREFS'} }, @{ $dependent_xrefs };
+  }
+
+  # FIXME: fix BaseParser to make dep-xref insertion via the graph replay-safe!!!
+
+  return $xref_graph_node;
+}
+
+
+# FIXME: description
+sub get_source_id_map {
+  my ( $self ) = @_;
+
+  # Just in case, even though we presently call _load_maps() in the
+  # constructor so it shouldn't be possible to call this method before
+  # maps have been loaded
+  if ( ! exists $self->{'maps'}->{'named_source_ids'} ) {
+    croak 'Source-ID map is missing';
+  }
+
+  return $self->{'maps'}->{'named_source_ids'};
+}
+
+
+sub _load_maps {
+  my ( $self, $baseParserInstance ) = @_;
+
+  my $taxonomy_ids_for_species
+    = $baseParserInstance->get_taxonomy_from_species_id( $self->{'species_id'},
+                                                         $self->{'dbh'} );
+  # If the map is empty, something is wrong
+  if ( scalar keys %{ $taxonomy_ids_for_species } == 0 ) {
+    croak "Got zero taxonomy_ids for species_id '"
+      . $self->{'species_id'} . q{'};
+  }
+  $self->{'maps'}->{'taxonomy_ids_for_species'}
+    = $taxonomy_ids_for_species;
+
+  my $source_id_map
+    = {
+       'Uniprot/SWISSPROT'
+       => {
+           'direct'          => undef,
+           'sequence_mapped' => undef,
+         },
+       'Uniprot/SPTREMBL'
+       => {
+           'direct'            => undef,
+           "protein_evidence_gt_$MAX_TREMBL_EVIDENCE_LEVEL_FOR_STANDARD" => undef,
+           'sequence_mapped'   => undef,
+         },
+     };
+  while ( my ( $source_name, $pri_ref ) = each %{ $source_id_map } ) {
+    foreach my $priority ( keys %{ $pri_ref } ) {
+      $pri_ref->{$priority}
+        = $baseParserInstance->get_source_id_for_source_name( $source_name,
+                                                              $priority,
+                                                              $self->{'dbh'} );
+      if ( $pri_ref->{$priority} == $ERR_SOURCE_ID_NOT_FOUND ) {
+        croak "No source ID found for source $source_name, priority $priority";
+      }
+    }
+  }
+  $self->{'maps'}->{'named_source_ids'}
+    = $source_id_map;
+
+  my %dependent_source_map
+    = $baseParserInstance->get_xref_sources( $self->{'dbh'} );
+  $self->{'maps'}->{'dependent_sources'}
+    = \%dependent_source_map;
+
+  return;
+}
+
+
+# Returns true if the current record describes an entry derived from
+# Ensembl, false otherwise.
+sub _entry_is_from_ensembl {
+  my ( $self ) = @_;
+
+  # As of end of October 2018, the old parser's way of identifying
+  # proteins derived from Ensembl by searching for a comment topic
```

Comment:
@magaliruffier - does this sound right to you?

Reply:
This is what Mag told me last week after having checked.

Reply:
Correct